sotodlib/io/bookbinder.py

```diff
@@ -1125,3 +1196,105 @@ def __call__(self):
 class TimingSystemError(Exception):
     """Exception raised when the timing system is on but timing counters are not found"""
     pass
+
+
+class HWPLoader:
+    """
+    Helper class to load in HWP data from data archive.
+    """
+    def __init__(self, root_dir):
+        self.root_dir = root_dir
+        self.files = []
+        self.data = {}
+        self.times = {} 
+
+        self.hwp_angle_interp = None
+        self.locked_interp = None
+        self.stable_interp = None
+        self.hwp_rate_interp = None
+
+    def get_files(self, start, stop):
+        """
+        Gets HWP files that may have data between start and stop time
+        """
+        files = []
+        for subdir in os.listdir(self.root_dir):
+            try:
+                tcode = int(subdir)
+            except:
+                continue
+
+            if not (((start//1e5) - 1) <= tcode <= stop//1e5):
+                continue
+
+            subpath = os.path.join(self.root_dir, subdir)
+            files.extend([os.path.join(subpath, f) for f in os.listdir(subpath)])
+
+        files = np.array(sorted(files))
+        file_times = np.array([int(os.path.basename(f).split('.')[0]) for f in files])
+
+        m = (start <= file_times) & (file_times < stop)
+        if not np.any(m):
+            self.files = []
+            return self.files
+
+        # Add file before for good measure
+        i0 = np.where(m)[0][0]
+        if i0 > 0:
+            m[i0 - 1] = 1
+
+        self.files = files[m].tolist()
+        return self.files
+
+    def frame_iter(self):
+        """Iterator that loops through G3 Frames of all files in self.files"""
+        return itertools.chain(*[core.G3File(f) for f in self.files])
+
+    def load_data(self, start, stop):
+        self.get_files(start, stop)
+
+        data = {}  # structured data[block_name][field]
+        times = {}  # structured times[block_name]
+
+        for frame in self.frame_iter():
+            if frame['hkagg_type'] != so3g.HKFrameType.data:
+                continue
+
+            for bname, block in zip(frame['block_names'], frame['blocks']):
+                if bname not in data:
+                    data[bname] = {
+                        k: [] for k in block.keys()
+                    }
+                    times[bname] = []
+
+                ts = np.array(block.times) / core.G3Units.s
+
+                # Includes a sec before and after for interpolation
+                m = ((start - 1) < ts) & (ts < (stop + 1))
```

Comment:
We should be careful about the assumption here. The start and stop times in bookbinder come from g3tsmurf db. @kmharrington 
could clarify how that timestamp is obtained. These timestamps often do not line up with the actual .times field, probably more so when timing system is on. Here it's assuming they can't be off by more than 1s. Is that a safe assumption?

Reply:
Ah this is a really good point. Since there is this uncertainty, I think we may as well just use a much larger buffer, of like `30 sec` or so on each side.